### ISSUE TITLE
Provide complete bindings for C11's libc

### ISSUE THREAD

**Issue by densh** [scala-native/scala-native#104]
- [ ] assert.h
- [ ] complex.h
- [ ] ctype.h
- [ ] errno.h
- [ ] fenv.h
- [ ] float.h
- [ ] inttypes.h
- [ ] iso646.h
- [ ] limits.h
- [ ] locale.h
- [ ] math.h
- [ ] setjmp.h
- [ ] signal.h
- [ ] stdalign.h
- [ ] stdarg.h
- [ ] stdatomic.h
- [ ] stdbool.h
- [ ] stddef.h
- [ ] stdint.h
- [ ] stdio.h
- [ ] stdlib.h
- [ ] stdnoreturn.h
- [ ] string.h
- [ ] tgmath.h
- [ ] threads.h
- [ ] time.h
- [ ] uchar.h
- [ ] wchar.h
- [ ] wctype.h

**Comment by densh** [comment]
That's a nice starting task for someone who wants to contribute some work. Just pick one of the header, go through all the functions and write signatures for them in terms of our native types that are defined in https://github.com/scala-native/scala-native/tree/master/nativelib/src/main/scala/scala/scalanative/native

**Comment by TimothyKlim** [comment]
I'll try to finish this task until the end of the week.

**Comment by mkotsbak** [comment]
Isn't this possible to automate using a script?

**Comment by TimothyKlim** [comment]
I'm using own python + clang script to do this: https://gist.github.com/TimothyKlim/a65289901221451e515e408014eb3613

**Comment by densh** [comment]
@TimothyKlim This looks great, would you be interested in expanding this script into full-blown tool for automatic binding generation? We'd love to home something like this in the project.

**Comment by TimothyKlim** [comment]
@densh I can't promise but I'll try to complete this as a tool within this task.

**Comment by TimothyKlim** [comment]
Which convention I need to use to declaring necessary constants in extern object? For example, I have `signal.h` constants as macros which cannot be extracted:
```c
#define	SIGHUP	1	/* hangup */
#define	SIGINT	2	/* interrupt */
#define	SIGQUIT	3	/* quit */
//...
```
and I can't declare these as `final val`s in extern object.

**Comment by mkotsbak** [comment]
@TimothyKlim ah nice. Then you just need to run it on libclang to convert the script to Scala native :)

**Comment by mkotsbak** [comment]
Hmm, regarding constants, maybe use Enumeration instead?

**Comment by sjrd** [comment]
> Which convention I need to use to declaring necessary constants in extern object?

In Scala.js, for similar situations, we declare them as `final val`s in a separate non-extern object.

**Comment by TimothyKlim** [comment]
>In Scala.js, for similar situations, we declare them as final vals in a separate non-extern object.

@sjrd, I saw that scala.js supports `@native trait`s and I can easily extend an object with my `val`s (in scala native I can't do it right now). Which naming format are you using for separate object?

**Comment by sjrd** [comment]
In Scala.js, an `@js.native trait` can only be extended by an `@js.native class`, so the vals in the extended class must also be native. I don't think this is relevant in this context.

> Which naming format are you using for separate object?

There's no strict rule. In most cases JS APIs appear in classes/traits more than objects, which allows us to simply use the companion object of the native class/trait to store non-native constants. This doesn't work in Scala Native, because most (all?) externs are already objects. In this case I would simply suggest a meaningful name such as
```scala
object Signal {
  final val SIGHUP = 1
  ...
}
```

**Comment by TimothyKlim** [comment]
Got it, thanks!

**Comment by densh** [comment]
@sjrd's advice is spot on.

**Comment by densh** [comment]
@TimothyKlim Are you still working on this?

**Comment by TimothyKlim** [comment]
@densh, I'm currently don't have much time to complete some headers and publish python script as a tool. I'll try to find free time on this week to get done some of these headers if no one will take it.

**Comment by densh** [comment]
As @pbatko discovered `assert.h` is out of scope for this one. Probably there are other headers which do not make much sense for us.

**Comment by jonas** [comment]
Regarding the ability to generate API binding from C header files, how hard would it be to put such a thing into the sbt plugin? It would require a small binding to clang's C API, the ability to compile/bootstrap those bindings and a say to run the scala native program using an SBT task with C include statements or header file path as the input setting.

**Comment by densh** [comment]
@jonas Integrating a binding generation tool into sbt plugin is doable. We just need someone to develop the tool itself. @frgomes has been exploring this in https://github.com/frgomes/scala-bindgen

**Comment by mkotsbak** [comment]
Yes, I would say the rules for making a facade for a C api that provides header files is a certain science, so I guess sometime in the future building a Scala Native program should be no more work than building a C application that uses other libraries (by just pointing on the header file). Wrapper objects should be generated automatically.

This is not comparable to Scala.Js, where the types are not known and might need checking documentation, examples or implementation to discover of what types the parameters are. That said, it could make sense to make a Scala wrapper on top of that again to get a more Scala feeling API, but that could come in addition where it makes sense.

**Comment by MasseGuillaume** [comment]
[time.h](https://github.com/scala-native/scala-native/pull/530/files#diff-0198121f75bd33074ba7ef926d77ba9eR1) is in a PR.

**Comment by MasseGuillaume** [comment]
I would uncheck errno.h since all values are not available:

from http://man7.org/linux/man-pages/man3/errno.3.html

```
EPERM
ENOENT
ESRCH
EINTR
EIO
ENXIO
E2BIG
ENOEXEC
EBADF
ECHILD
EAGAIN
ENOMEM
EACCES
EFAULT
ENOTBLK
EBUSY
EEXIST
EXDEV
ENODEV
ENOTDIR
EISDIR
EINVAL
ENFILE
EMFILE
ENOTTY
ETXTBSY
EFBIG
ENOSPC
ESPIPE
EROFS
EMLINK
EPIPE
EDOM
ERANGE
```

**Comment by densh** [comment]
@MasseGuillaume POSIX and libc are not the same thing. libc is way more restrictive and is available on all platforms, POSIX APIs on the other hand are not really there on windows.

**Comment by sorenbug** [comment]
I attempted stdbool mappings: #1499
